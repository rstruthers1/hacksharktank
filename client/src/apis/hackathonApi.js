// hackathonApi.js
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const hackathonApi = createApi({
    reducerPath: 'hackathonApi',
    baseQuery: fetchBaseQuery(
        {baseUrl: `${baseURL}/`}), // Set your base URL here
    endpoints: (builder) => ({
        createHackathon: builder.mutation({
            query: (hackathonData) => ({
                url: 'hackathons', // Your endpoint path
                method: 'POST',
                body: hackathonData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
        }),
        updateHackathon: builder.mutation({
            query: (hackathonData) => ({
                url: `hackathons/${hackathonData.id}`, // Your endpoint path
                method: 'PUT',
                body: hackathonData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'Hackathons'},
                { type: 'Hackathon', id: arg.id }],
        }),
        getHackathons: builder.query({
            query: () => ({
                url: 'hackathons', // Your endpoint path
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'Hackathons'}],
        }),
        getHackathon: builder.query({
            query: (hackathonId) => ({
                url: `hackathons/${hackathonId}`, // Your endpoint path
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'Hackathon', id: arg }],
        }),
        createHackathonUserRole: builder.mutation({
            query: (hackathonUserRoleData) => ({
                url: 'hackathons/users/roles', // Your endpoint path
                method: 'POST',
                body: hackathonUserRoleData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg.hackathonId }],
        }),
        deleteHackathonUserRole: builder.mutation({
            query: (hackathonUserRoleData) => ({
                url: `hackathons/${hackathonUserRoleData.hackathonId}/users/${hackathonUserRoleData.userId}/roles/${hackathonUserRoleData.roleName}`, // Your endpoint path
                method: 'DELETE',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg.hackathonId }],
        }),
        getHackathonUsers: builder.query({
            query: (hackathonId) => ({
                url: `hackathons/${hackathonId}/users`, // Your endpoint path
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg }],
        }),
        deleteHackathonUser: builder.mutation({
            query: (hackathonUserData) => ({
                url: `hackathons/${hackathonUserData.hackathonId}/users/${hackathonUserData.userId}`, // Your endpoint path
                method: 'DELETE',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg.hackathonId }],
        }),
        // You can add more endpoints here
    }),
});

export const {
    useCreateHackathonMutation,
    useUpdateHackathonMutation,
    useGetHackathonsQuery,
    useCreateHackathonUserRoleMutation,
    useGetHackathonQuery,
    useGetHackathonUsersQuery,
    useDeleteHackathonUserRoleMutation,
    useDeleteHackathonUserMutation,
} = hackathonApi;
