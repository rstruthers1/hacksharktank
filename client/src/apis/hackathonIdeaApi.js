import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const hackathonIdeaApi = createApi({
    reducerPath: 'hackathonIdeaApi',
    baseQuery: fetchBaseQuery(
        {baseUrl: `${baseURL}/`}), // Set your base URL here
    endpoints: (builder) => ({
        getHackathonIdeas: builder.query({
            query: (hackathonId) => ({
                url: `hackathons/${hackathonId}/ideas`,
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
        }),
       createHackathonIdea: builder.mutation({
                query: (hackathonIdeaData) => ({
                    url: `hackathons/${hackathonIdeaData.hackathonId}/ideas`,
                    method: 'POST',
                    body: hackathonIdeaData.data,
                    // Add a JWT token to the request headers if the user is logged in
                    headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
                }),
                invalidatesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
            }),
        updateHackathonIdea: builder.mutation({
            query: (hackathonIdeaData) => ({
                url: `hackathons/${hackathonIdeaData.hackathonId}/ideas`,
                method: 'PUT',
                body: hackathonIdeaData.data,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
        }),
        deleteHackathonIdea: builder.mutation({
            query: ({hackathonId, ideaId}) => ({
                url: `hackathons/${hackathonId}/ideas/${ideaId}`,
                method: 'DELETE',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
        }),
    })
});

export const {
    useGetHackathonIdeasQuery,
    useCreateHackathonIdeaMutation,
    useDeleteHackathonIdeaMutation,
    useUpdateHackathonIdeaMutation
} = hackathonIdeaApi;
