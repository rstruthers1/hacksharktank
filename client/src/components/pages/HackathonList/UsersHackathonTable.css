/* HackathonTable.css */
.idColumn {
    width: 5%;
}

.eventNameColumn {
    width: 20%;
}

.descriptionColumn {
    width: 35%;
}

.startDateColumn {
    width: 12%;
}

.endDateColumn {
    width: 12%;
}

.roleColumn {
    width: 16%;
}


/* HackathonTable.css */
.table-container {
    display: flex;
    flex-direction: column;
    height: 400px; /* Adjust based on your needs */
}

.table-scrollable {
    flex-grow: 1;
    overflow-y: auto;
}

.pagination-container {
    text-align: center; /* Center align pagination controls */
}

/* HackathonTable.css */
.search-container {
    margin-bottom: 20px; /* Adds space between the search box and the table */
    text-align: center; /* Centers the search box */
}

.search-container input {
    width: 50%; /* Adjust the width as per your layout */
    padding: 10px;
    margin: 0 auto;
    border: 1px solid #ddd; /* Subtle border */
    border-radius: 4px; /* Rounded corners */
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.1); /* Soft box shadow for depth */
}

.search-container input:focus {
    outline: none; /* Removes the default focus outline */
    border-color: #007bff; /* Highlight color when the input is focused */
    box-shadow: 0 0 5px rgba(0, 123, 255, 0.5); /* Matching shadow color */
}


