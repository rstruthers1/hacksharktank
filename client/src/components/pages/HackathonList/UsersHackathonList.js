import {Container} from "react-bootstrap";
import {useGetUsersHackathonsQuery} from "../../../apis/hackathonApi";
import {getLoggedInUser} from "../../../utils/authUtils";
import UsersHackathonTable from "./UsersHackathonTable";

const UsersHackathonList = () => {
    // Get hackathons from backend
    const {data: hackathons, error, isLoading} = useGetUsersHackathonsQuery(getLoggedInUser().id);

    return (
        <>
            <p>Hackathons that you are participating in:</p>
            <Container>
                {error ? (
                    <>Oh no, there was an error</>
                ) : isLoading ? (
                    <>Loading...</>
                ) : hackathons ? (
                    <UsersHackathonTable hackathons={hackathons}/>
                ) : null}
            </Container>
        </>
    )
}

export default UsersHackathonList;