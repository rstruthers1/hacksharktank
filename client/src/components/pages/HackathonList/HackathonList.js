import {Container} from "react-bootstrap";
import {useGetHackathonsQuery} from "../../../apis/hackathonApi";
import HackathonTable from "./HackathonTable";

const HackathonList = () => {
    // Get hackathons from backend
    const { data: hackathons, error, isLoading } = useGetHackathonsQuery();

    return (
        <Container>
            <h1>Hackathons</h1>
            {error ? (
                <>Oh no, there was an error</>
            ) : isLoading ? (
                <>Loading...</>
            ) : hackathons ? (
                <HackathonTable hackathons={hackathons}/>
            ) : null}
        </Container>
    )
}

export default HackathonList;