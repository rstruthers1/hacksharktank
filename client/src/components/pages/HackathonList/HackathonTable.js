import HackathonTableBase from './HackathonTableBase';
import {Link} from "react-router-dom";
import {formatDateTimeAsDate} from "../../../utils/dateTimeUtils";


const HackathonTable = ({ hackathons }) => {
    const columns = ['ID', 'Event Name', 'Description', 'Start Date', 'End Date'];
    const linkPath = '/admin/hackathon/';

    const renderRow = (hackathon) => (
        <tr key={hackathon.id}>
            <td>{hackathon.id}</td>
            <td><Link to={`${linkPath}${hackathon.id}/edit`}>{hackathon.eventName}</Link></td>
            <td>{hackathon.description}</td>
            <td>{formatDateTimeAsDate(hackathon.startDate)}</td>
            <td>{formatDateTimeAsDate(hackathon.endDate)}</td>
        </tr>
    );

    return <HackathonTableBase hackathons={hackathons} columns={columns} renderRow={renderRow} />;
};

export default HackathonTable;