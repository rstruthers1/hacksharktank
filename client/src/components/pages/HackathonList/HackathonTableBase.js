import React, { useState } from 'react';
import { Table, Pagination } from 'react-bootstrap';
import './HackathonTable.css';

const HackathonTableBase = ({ hackathons, itemsPerPage = 5, columns, renderRow }) => {
    const [activePage, setActivePage] = useState(1);
    const [searchTerm, setSearchTerm] = useState('');

    const filteredHackathons = searchTerm
        ? hackathons.filter(hackathon =>
            hackathon.eventName.toLowerCase().includes(searchTerm.toLowerCase())
        )
        : hackathons;

    const indexOfLastItem = activePage * itemsPerPage;
    const indexOfFirstItem = indexOfLastItem - itemsPerPage;
    const currentItems = filteredHackathons.slice(indexOfFirstItem, indexOfLastItem);

    const paginate = (pageNumber) => setActivePage(pageNumber);

    const totalPages = Math.ceil(hackathons.length / itemsPerPage);
    let items = [];
    for (let number = 1; number <= totalPages; number++) {
        items.push(
            <Pagination.Item key={number} active={number === activePage} onClick={() => paginate(number)}>
                {number}
            </Pagination.Item>,
        );
    }

    return (
        <div className="table-container">
            <div className="search-container">
                <input
                    type="text"
                    placeholder="Search by event name..."
                    onChange={(e) => setSearchTerm(e.target.value)}
                />
            </div>
            <div className="table-scrollable">
                <Table bordered hover>
                    <thead>
                    <tr>
                        {columns.map(column => (
                            <th key={column} className={`${column}Column`}>{column}</th>
                        ))}
                    </tr>
                    </thead>
                    <tbody>
                    {currentItems.map((hackathon) => (
                        renderRow(hackathon)
                    ))}
                    </tbody>
                </Table>
            </div>
            <div className="pagination-container">
                <Pagination>
                    <Pagination.First onClick={() => paginate(1)} disabled={activePage === 1} />
                    <Pagination.Prev onClick={() => paginate(activePage - 1)} disabled={activePage === 1} />
                    {items}
                    <Pagination.Next onClick={() => paginate(activePage + 1)} disabled={activePage === totalPages} />
                    <Pagination.Last onClick={() => paginate(totalPages)} disabled={activePage === totalPages} />
                </Pagination>
            </div>
        </div>
    );
};

export default HackathonTableBase;