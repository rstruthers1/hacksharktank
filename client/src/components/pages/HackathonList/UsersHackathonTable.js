import HackathonTableBase from './HackathonTableBase';
import {Link} from "react-router-dom";
import {formatDateTimeAsDate} from "../../../utils/dateTimeUtils";

const UsersHackathonTable = ({ hackathons }) => {
    const columns = ['ID', 'Event Name', 'Description', 'Start Date', 'End Date', 'Your Roles'];
    const linkPath = '/dashboard/hackathon/';

    const renderRow = (hackathon) => (
        <tr key={hackathon.id}>
            <td>{hackathon.id}</td>
            <td><Link to={`${linkPath}${hackathon.id}/about`}>{hackathon.eventName}</Link></td>
            <td>{hackathon.description}</td>
            <td>{formatDateTimeAsDate(hackathon.startDate)}</td>
            <td>{formatDateTimeAsDate(hackathon.endDate)}</td>
            <td>{hackathon.hackathonUserRoles.map((role) => (
                <div key={role}>{role}</div>
            ))}</td>
        </tr>
    );

    return <HackathonTableBase hackathons={hackathons} columns={columns} renderRow={renderRow} />;
};

export default UsersHackathonTable;