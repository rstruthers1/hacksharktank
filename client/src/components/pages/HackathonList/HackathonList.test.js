import React from 'react';
import { render, screen } from '@testing-library/react';
import HackathonList from './HackathonList';
import { useGetHackathonsQuery } from '../../../apis/hackathonApi';
import {createMemoryHistory} from "history";
import {Router} from "react-router-dom";
import renderer from "react-test-renderer";
import 'cross-fetch/polyfill';
import '@testing-library/jest-dom';

jest.mock('../../../apis/hackathonApi');

describe('HackathonList', () => {

    it('renders correctly', () => {
        const history = createMemoryHistory();
        const mockHackathons = [
            { id: 1, eventName: 'Test Hackathon 1', description: 'Description 1', startDate: '2022-01-01', endDate: '2022-01-02' },
            { id: 2, eventName: 'Test Hackathon 2', description: 'Description 2', startDate: '2022-02-01', endDate: '2022-02-02' },
        ];
        useGetHackathonsQuery.mockReturnValue({ data: mockHackathons, isSuccess: true });
        const tree = renderer.create(<Router location={history.location} navigator={history}><HackathonList /></Router>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders loading state', () => {
        useGetHackathonsQuery.mockReturnValue({ isLoading: true });
        render(<HackathonList />);
        expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('renders error state', () => {
        useGetHackathonsQuery.mockReturnValue({ error: new Error() });
        render(<HackathonList />);
        expect(screen.getByText('Oh no, there was an error')).toBeInTheDocument();
    });

    it('renders data', () => {
        const history = createMemoryHistory();
        const mockHackathons = [
            { id: 1, eventName: 'Test Hackathon 1', description: 'Description 1', startDate: '2022-01-01', endDate: '2022-01-02' },
            { id: 2, eventName: 'Test Hackathon 2', description: 'Description 2', startDate: '2022-02-01', endDate: '2022-02-02' },
        ];
        useGetHackathonsQuery.mockReturnValue({ data: mockHackathons, isSuccess: true });
        render(<Router location={history.location} navigator={history}><HackathonList /></Router>);
        expect(screen.getByText('Hackathons')).toBeInTheDocument();
    });
});