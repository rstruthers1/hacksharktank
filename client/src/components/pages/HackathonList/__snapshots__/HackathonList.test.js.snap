// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackathonList renders correctly 1`] = `
<div
  className="container"
>
  <h1>
    Hackathons
  </h1>
  <div
    className="table-container"
  >
    <div
      className="search-container"
    >
      <input
        onChange={[Function]}
        placeholder="Search by event name..."
        type="text"
      />
    </div>
    <div
      className="table-scrollable"
    >
      <table
        className="table table-bordered table-hover"
      >
        <thead>
          <tr>
            <th
              className="IDColumn"
            >
              ID
            </th>
            <th
              className="Event NameColumn"
            >
              Event Name
            </th>
            <th
              className="DescriptionColumn"
            >
              Description
            </th>
            <th
              className="Start DateColumn"
            >
              Start Date
            </th>
            <th
              className="End DateColumn"
            >
              End Date
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              1
            </td>
            <td>
              <a
                href="/admin/hackathon/1/edit"
                onClick={[Function]}
              >
                Test Hackathon 1
              </a>
            </td>
            <td>
              Description 1
            </td>
            <td>
              12/31/2021
            </td>
            <td>
              1/1/2022
            </td>
          </tr>
          <tr>
            <td>
              2
            </td>
            <td>
              <a
                href="/admin/hackathon/2/edit"
                onClick={[Function]}
              >
                Test Hackathon 2
              </a>
            </td>
            <td>
              Description 2
            </td>
            <td>
              1/31/2022
            </td>
            <td>
              2/1/2022
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      className="pagination-container"
    >
      <ul
        className="pagination"
      >
        <li
          className="page-item disabled"
        >
          <span
            className="page-link"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              «
            </span>
            <span
              className="visually-hidden"
            >
              First
            </span>
          </span>
        </li>
        <li
          className="page-item disabled"
        >
          <span
            className="page-link"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              ‹
            </span>
            <span
              className="visually-hidden"
            >
              Previous
            </span>
          </span>
        </li>
        <li
          className="page-item active"
        >
          <span
            className="page-link"
            onClick={[Function]}
          >
            1
            <span
              className="visually-hidden"
            >
              (current)
            </span>
          </span>
        </li>
        <li
          className="page-item disabled"
        >
          <span
            className="page-link"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              ›
            </span>
            <span
              className="visually-hidden"
            >
              Next
            </span>
          </span>
        </li>
        <li
          className="page-item disabled"
        >
          <span
            className="page-link"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              »
            </span>
            <span
              className="visually-hidden"
            >
              Last
            </span>
          </span>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
